src/Remote/Importer.php

```diff
@@ -0,0 +1,162 @@
+            }
+
+            $this->importPackageWithComposer($package, version: $version ?? '*', repositoryUrl: $vcs, file: $file);
+
+            return;
+        }
+
+        if ('package' === $scheme) {
+            if (null !== $version || null !== $vcs) {
+                throw new InvalidImportFormat('The "source" and "vcs" arguments are not supported for non-Composer packages.');
+            }
+            if (null === $source) {
+                throw new InvalidImportFormat('The "source" argument is required for non-Composer packages.');
+            }
+
+            $this->importPackageWithComposer($package, version: 'v1', source: $source, file: $file);
+
+            return;
+        }
+
+        throw new InvalidImportFormat(sprintf('The import scheme "%s" is not supported.', $scheme));
+    }
+
+    public function fetchPackages(): void
+    {
+        if (!$this->imports) {
+            $this->composer->remove();
+
+            return;
+        }
+
+        $input = GlobalHelper::getApplication()->getInput();
+
+        // Need to look for the raw options as the input is not yet parsed
+        $forceUpdate = null === $input->getParameterOption('--update-remotes', true);
+        $displayProgress = null !== $input->getParameterOption('--no-interaction', true);
+
+        $autoloadPath = PathHelper::getRoot() . Composer::VENDOR_DIR . 'autoload.php';
+
+        if (!file_exists($autoloadPath)) {
+            $forceUpdate = true;
+        }
+
+        $this->composer->update($forceUpdate, $displayProgress);
+
+        require_once $autoloadPath;
+
+        foreach ($this->imports as $package => $data) {
+            foreach ($data as $version => $files) {
+                foreach ($files as $file) {
+                    import(PathHelper::getRoot() . Composer::VENDOR_DIR . $package . '/' . ($file ?? ''));
+                }
+            }
+        }
+    }
+
+    /**
+     * @param ?array{
+     *     url?: string,
+     *     type?: "git" | "svn",
+     *     reference?: string,
+     * } $source
+     */
+    private function importPackageWithComposer(string $package, string $version, ?string $repositoryUrl = null, ?array $source = null, ?string $file = null): void
+    {
+        $this->logger->notice('Importing remote package with Composer.', [
+            'package' => $package,
+            'version' => $version,
+        ]);
+
+        $json = $this->composer->getConfiguration();
+
+        $json['require'][$package] = $version;
+
+        if ($repositoryUrl) {
+            $json['repositories'][] = [
+                'type' => 'vcs',
+                'url' => $repositoryUrl,
+            ];
+        }
+
+        if ($source) {
+            if (!isset($source['url'], $source['type'], $source['reference'])) {
+                throw new ImportError('The "source" argument must contain "url", "type" and "reference" keys.');
```

Comment:
Is this dead code? because eveything has been check before?

* if yes, you can resolve this comment
* if not, could you add more information in the exception, to help the user. (name of the packages, etc)

Reply:
Not sure to understand, `$source` structure is never checked

Reply:
About the more information, it will be done higher, on the import function when the error is catched